docs/rearrange.md

```diff
@@ -0,0 +1,180 @@
+
+# unordered syntax allows for select dimensions by name
+# transpose last two dimensions
+y = hax.rearrange(x, "{H W} -> ... W H")
+
+# don't know the order? use the unordered syntax
+y = hax.rearrange(x, "{W C H N} -> N H W C")
+
+# split dims as in einops
+y = hax.rearrange(x, "(step microbatch) ... -> step microbatch ...", step=4)
+# splitting dims can be done w/ unordered syntax
+y = hax.rearrange(x, "{(N: step microbatch) ...} -> step microbatch ...", step=4)
+
+# merging dims requires a name
+x = hax.rearrange(y, "step microbatch ... -> (N: step microbatch) ...")
+
+# some fancier examples
+
+# split into patches
+y = hax.rearrange(x, "N C (patch_h H) (patch_w W) -> N (P: patch_h patch_w) C H W", H=4, W=4)
+# unordered version
+y = hax.rearrange(x, "{(H: patch_h H) (W: patch_w W) C } -> ... (P: patch_h patch_w) C H W", H=4, W=4)
+
+# split into patches, then merge patches and channels
+y = hax.rearrange(x, "N C (patch_h H) (patch_w W) -> N (P: patch_h patch_w) (C: C H W)", H=4, W=4)
+# unordered version
+y = hax.rearrange(x, "{(H: patch_h H) (W: patch_w W) C } -> ... (P: patch_h patch_w) (C: C H W)", H=4, W=4)
+```
+
+### Bindings: Aliasing and Sizing
+
+In the above examples we used keyword arguments to give sizes to split axes, which is the same
+as in einops. However, we can also use bindings to alias axes. For example:
+
+```python
+# this produces the same result as the previous example
+y2 = hax.rearrange(x, "N C (patch_h foo) (patch_w bar) -> N (P: patch_h patch_w) (C: C foo bar)", foo=hax.Axis("H", 4), bar=hax.Axis("W", 4))
+assert y.axes == y2.axes
+```
+
+This example is a bit contrived, but it lets us use shorter or different names in the string which is occasionally useful.
+You can actually just use strings to specify aliases, which is sometimes useful as well.
+For instance, if we wanted "P" to actually be called "patch", but wanted to keep the short syntax, we could do:
+
+```python
+y3 = hax.rearrange(x, "N C (nh ph) (nw pw) -> N (P: nh nw) (C: C ph pw)", P="patch", pw=4, ph=4)
+```
+
+
+
+### Differences from einops
+
+As you may have noticed, there are some differences from einops:
+
+* Merged axes must have a name: `(C: C H W)` instead of `(C H W)`
+* The unordered syntax with `{ ... }` is new but hopefully not too confusing: select dimensions by name, and optionally split them using, e.g., `(N: step microbatch)`.
+* As discussed immediately above, you can use bindings to specify axis objects for names as well as sizes.
+
+### Syntax
+
+Semiformally, the syntax is an lhs and an rhs. The lhs is either ordered or unordered:
+
+* *ordered lhs*: a sequence of axis variable names or (named or anonymous) split axes, separated by spaces or commas, and up to one ellipsis
+* *unordered lhs*: a sequence of axis names or named split axes, surrounded by `{}`, separated by spaces or commas
+
+* An *axis variable name* is an identifier (that need not correspond to an axis name in the input or output.)
```

Comment:
```suggestion
* An *axis variable name* is an identifier (that does not need to correspond to an axis name in the input or output.)
```

Reply:
I don't fully understand the difference between axis variable name vs. axis name here. 
In the example of `N = hax.Axis("N", 32)`, is `"N"` the axis name and `N` the axis variable name?

Reply:
added a note. that point is that the string in an axis variable is arbitrary, while an axis name must refer to an actual axis